design/StableRegions-20230209.md

```diff
@@ -0,0 +1,218 @@
+(256GB compared to 32GB, the limit today)
+
+### Q: When is stable memory becoming less costly?
+
+TBA
+
+### Q: How does the cheaper stable memory access cost compare with ordinary heap memory access cost?
+
+TBA
+
+
+## Design details
+
+### API
+
+Internal region allocator operations:
+
+
+ - `initialize` -- called by the RTS, not by the Motoko developer.
+ - `rebuild` -- also called by the RTS, during an post-upgrade.
+
+
+User-facing region allocator operations:
+
+ - `region_new` -- create a dynamic region.
+ - `region_grow` -- grow region by a specified number of pages.
+ - `region_load` -- read some data from the region.
+ - `region_store` -- store some data into the region.
+
+And a special operation, for testing our design for future GC integration (bonus):
+
+- `region_release` -- release region and reuse region's page blocks.
+
+The `_release` operation is *not* part of the user-facing API nor part of the MVP,
+but supporting it is important because it means we can transition quickly to an integration
+with the ambient Motoko GC if we can support it.
+
+## Internal footprint
+
+The state of the allocator is stored in a combination of stable memory and heap memory.
+
+The stable memory state is sufficient to reconstitute the heap objects
+(`rebuild` operation, described in a subsection below).  That means
+that even if the heap is lost (like during an upgrade that fails) the
+stable memory state can fully describe the parts that should be
+rebuilt into the heap when the upgrade succeeds.
+
+### stable memory fields
+
+ - total allocated blocks, `Nat16`, max value is `32768`.
+ - total allocated regions, `Nat16`, max value is `32767` (one region is reserved for "no region").
+ - The `block-region` table (fixed size, about 131kb).
+ - The `region-blocks` table (fixed size, about 524kb).
+
+### representation of values of type `Region`
+
+ - We use the region ID, a `Nat16`, to represent the region as a value.
+ - To compute a load or store access, this requires consulting stable memory for (almost all) meta data.
+ - We rely on the cost of these stable memory accesses going down in the near future to support this design decision.
+
+### block-region table
+
+ - purpose:
+   - relate a block ID ("page block ID") to its region (if any) and its position in that region (see `rebuild`).
+
+   - NB: The organization of this table is not useful for efficient
+     access calculations of load or store (they require a linear
+     search that would be prohibitively slow). Rather, the
+     `region-blocks` table is designed for that purpose.  OTOH, this
+     table is suitable to do a "batch" rebuild of the dynamic heap-allocated vectors
+     in that table, which get lost during upgrades.
+
+ - 32768 entries (statically sized).
+ - 4 bytes per entry.
+ - entry type = `BlockRegion { region : Nat16, position : Nat16 }`
+ - the location of each entry gives its corresponding block ID.
+
+### region-blocks table
```

Comment:
@luc-blaeser @crusso Can you double-check my rationale in this section, particularly the "NB" comments?

Reply:
I think this design is really good, allowing memory-economic and runtime-efficient runtime support, yet simple. Also the reconstruction of heap-based structures for fast access seems a good choice.
I just had a few thoughts:
* How would reclamation of page blocks be managed? A possibility could be to a keep a simple (linearly=linked) free list inside the `block-regions` table. Allocation and free would be `O(1)`, since all allocated page blocks are equally sized. 
* For the region-block table (that reside in normal heap), we could maybe use `Buffer`.

Reply:
> How would reclamation of page blocks be managed? 

I had thought that we'd reserve two special regions for dedicated purposes (region 0 and 1):
 - Region 0 would support the existing stable memory API, for an "anonymous", shared, global region.
 - Region 1 would be the reclaimed page blocks of other regions.  Push reclaimed blocks onto the end, and pop blocks to allocate again from the end, like a stack.
 
 WDYT?

Reply:
That sounds sensible to me, but I think you still need a way to indicate which region ID's are free, especially once the GC reclaims unusable ones (and their blocks). Or maybe I'm missing something.